### PR TITLE
Propose text search in attachments

### PR DESCRIPTION
**Pull request by chibenwa**
By the following proposal, I would like to submit a small addition to the specification to allow text search in attachments.

Attachments are the only non-searchable field, and I can hardly see why the specification should be restrictive on this.

You will notice the **may** about text extraction. It allows a default implementation that just index attachments with **text/\*** making a minimal implementation server side working in a question of minutes.

### CODE REVIEW

**Comment by cjnewman** [comment]
I support this proposal.

**Comment by chibenwa** [comment]
@neilj  or others, any opinions on this proposal?

File: spec/mail/searchsnippet.mdown [hunk 1]
```diff
@@ -10,8 +10,10 @@ A **SearchSnippet** object has the following properties:
   If text from the filter matches the subject, this is the subject of the message HTML-escaped, with matching words/phrases wrapped in `<mark></mark>` tags. If it does not match, this is `null`.
 - **preview**: `String|null`
   If text from the filter matches the plain-text or HTML body, this is the relevant section of the body (converted to plain text if originally HTML), HTML-escaped, with matching words/phrases wrapped in `<mark></mark>` tags, up to 256 characters long. If it does not match, this is `null`.
+- **attachments**: `String|null`
+  If text from the filter matches the text extracted from an attachment, this is the relevant section of the body (converted to plain text), with matching words/phrases wrapped in `<mark></mark>` tags, up to 256 characters long. If it does not match, this is `null`.
```
Comment (neilj) [inline spec/mail/searchsnippet.mdown]: "body" => "attachment"

File: spec/mail/searchsnippet.mdown [hunk 2]
```diff
@@ -10,8 +10,10 @@ A **SearchSnippet** object has the following properties:
...
 
-It is server-defined what is a relevant section of the body for preview. If the server is unable to determine search snippets, it MUST return `null` for both the *subject* and *preview* properties.
+It is server-defined what is a relevant section of the body for preview. If the server is unable to determine search snippets, it MUST return `null` for both the *subject*, *preview* and *attachment* properties.
```
Comment (neilj) [inline spec/mail/searchsnippet.mdown]: attachment => attachments

File: spec/mail/messagelist.mdown [hunk 3]
```diff
@@ -89,6 +89,8 @@ A **FilterCondition** object has the following properties:
   Looks for the text in the *subject* property of the message.
 - **body**: `String|null`
   Looks for the text in the *textBody* or *htmlBody* property of the message.
+- **attachments**: `String|null`
+  Looks for the text in the attachments of the message. Server MAY handle text extraction when possible for the different kinds of media.
```
Comment (neilj) [inline spec/mail/messagelist.mdown]: This should probably be searched automatically if you search with the "text" property too (or maybe we redefine that to simply be more generic, so we don't have to explicitly list all the others it matches).

Comment (chibenwa) [reply spec/mail/messagelist.mdown]: Actually I believe having the semantic available to allow explicitly looking up in attachments is a bonus. It allows to add such features like:

<<Oh, I need to retrieve the email with the odt document about last strategy meeting>>

Is clearly more specific than:

<<Give me all the mail about strategy meeting>>

The idea you gave (weakening text search) actually can be emulated with a AND between text and attachment search.

More generally, I would value having searches close to the message object.

Comment (chibenwa) [reply spec/mail/messagelist.mdown]: Is clearly more specific than:

Search all the emails that contains 'strategy meeting'

Sorry for the display issue.

Comment (neilj) [reply spec/mail/messagelist.mdown]: No, you misunderstood my comment. I'm not suggesting removing the *attachments* property for explicitly searching within attachments.

The *text* property is a shorthand for searching all the text-based properties (to, body etc.), and so I'm saying as *attachments* is a text-based search, this should also be referenced in the list of properties that *text* searches, or probably better update the definition of *text* so it automatically includes all text-based properties added in the future.

Comment (cjnewman) [reply spec/mail/messagelist.mdown]: Q for Neil: do you want 'text' search to cover all future non-mutable text-based searches, or all future text based searches. This detail matters when annotation support (RFC 5257) is added.

**Comment by timofonic** [comment]
I agree. This is a nice addition 👍

File: spec/mail/messagelist.mdown (hunk 3 above)
Comment (chibenwa) [reply spec/mail/messagelist.mdown]: Ok, yes, sure. I will update the proposal to match this.

Comment (neilj) [reply spec/mail/messagelist.mdown]: cjnewman: Good question. I think annotations should probably always be explicit, as whether they make sense to include depends on the semantics the annotation is being used for.

**Pull request merged**